### PR TITLE
Add feature Activation Bias Correction

### PR DESCRIPTION
**Pull request by lapid92**
## Pull Request Description:
Activation Bias Correction feature calculates the quantization error from activation quantization and "corrects" it in the next layer's bias.
The feature is set to False by default.

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    Calculate the centers of bins given their edges.
+
+    Parameters:
+    bin_edges: Array of bin edges.
+
+    Returns:
+    np.ndarray: Array of bin centers.
+    """
+    # Ensure bin_edges is a numpy array
+    bin_edges = np.array(bin_edges, dtype=np.float32)
```
Comment (elad-c) [inline model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py:82]: if you're not sure of the argument type, state all type options in the type hints.
if you are sure it's numpy array, change to assert

File: model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        prev_node: Node to compute the activation error caused by his activation quantization.
+        kernel_size: The framework specific attribute name of the convolution layer's kernel size.
+
+    Returns:
+        Graph with activation bias correction term for each node.
+    """
+
+    # Check if 'kernel_size' is a key in the framework-specific attributes of the linear_node, if it is then the
+    # linear_node is a convolution
+    if kernel_size in linear_node.framework_attr.keys():
```
Comment (elad-c) [inline model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py:115]: this `if` is redundant

File: model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        # Retrieve the value of 'kernel_size' and check if it is not 1 or (1, 1). This feature supports only kernel
+        # size of 1 or (1, 1).
+        if linear_node.framework_attr.get(kernel_size) not in [1, (1, 1)]:
+            # If the kernel size is not 1 or (1, 1), return the current graph unmodified
+            return graph
+
+    prev_node_act_quant_cfg = prev_node.final_activation_quantization_cfg
+
+    # Check if the previous node's has activation quantization configuration and if the previous node have the
+    # histogram collector
```
Comment (elad-c) [inline model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py:125]: add . at the end of comments.

File: model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    # Calculate the mean of the both the float and the quantized bin centers, weighted by the bin counts
+    mean_float_centers = np.sum(float_centers * float_count) / np.sum(float_count)
+    mean_quant_centers = np.sum(quant_centers * float_count) / np.sum(float_count)
+
+    # Compute the difference between the mean quantized center and the mean float center
+    mean_diff = mean_quant_centers - mean_float_centers
+
+    # Check if activation bias correction is enabled based on the configured threshold
+    if core_config.quantization_config.activation_bias_correction_threshold > 0:
```
Comment (elad-c) [inline model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py:147]: this seems wrong. init the threshold to inf if you want it to be enabled by default.

File: model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+        # Calculate the normalized bias as a percentage of the float center norm
+        float_centers_norm1 = np.abs(mean_float_centers)
+        normalized_bias = 100 * np.abs(mean_diff) / float_centers_norm1
+
+        # If the normalized bias is below the activation bias correction threshold, return the unmodified graph
+        if normalized_bias < core_config.quantization_config.activation_bias_correction_threshold:
+            return graph
+
+    # The correction term is a function of the layer type.
```
Comment (elad-c) [inline model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py:157]: what?

File: model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    kernel = linear_node.get_weights_by_keys(fw_info.kernel_ops_attributes_mapping.get(linear_node.type)[0])
+
+    if kernel is not None:
+        output_channel_index, input_channel_index = fw_info.kernel_channels_mapping.get(linear_node.type)
```
Comment (elad-c) [inline model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py:161]: explain all this logic

File: model_compression_toolkit/core/keras/keras_implementation.py [hunk 7]
```diff
@@ -218,6 +220,25 @@ def shift_negative_correction(self,
...
+    def compute_activation_bias_correction(self,
+                                           graph: Graph,
+                                           core_config: CoreConfig,
+                                           fw_info: FrameworkInfo):
+        """
+        Compute activation bias correction on a graph.
+
+        Args:
+            graph: Graph to apply activation bias correction on.
+            core_config: QuantizationConfig of how the model should be quantized.
```
Comment (elad-c) [inline model_compression_toolkit/core/keras/keras_implementation.py:232]: pass only the `QuantizationConfig` if you don't need the entire `CoreConfig`.

File: model_compression_toolkit/core/keras/statistics_correction/keras_compute_activation_bias_correction_of_graph.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+from model_compression_toolkit.core.common.framework_info import FrameworkInfo
+from model_compression_toolkit.core.common import Graph
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher
+from model_compression_toolkit.core.common.statistics_correction.compute_activation_bias_correction_of_graph import \
+    compute_activation_bias_correction_of_graph
+
+
+def activation_bias_correction_node_matchers():
+    # Match linear layers where we can add a correction.
+    linear_node = NodeOperationMatcher(Conv2D) | \
```
Comment (elad-c) [inline model_compression_toolkit/core/keras/statistics_correction/keras_compute_activation_bias_correction_of_graph.py:39]: Conv2DTranspose?

File: model_compression_toolkit/core/pytorch/statistics_correction/pytorch_compute_activation_bias_correction_of_graph.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+from model_compression_toolkit.core.common.framework_info import FrameworkInfo
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher
+from model_compression_toolkit.core.common.statistics_correction.compute_activation_bias_correction_of_graph import \
+    compute_activation_bias_correction_of_graph
+from model_compression_toolkit.core.pytorch.constants import KERNEL_SIZE
+
+
+def activation_bias_correction_node_matchers():
+    # Match linear layers where we can add a correction.
+    linear_node = NodeOperationMatcher(Linear) | NodeOperationMatcher(Conv2d)
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/statistics_correction/pytorch_compute_activation_bias_correction_of_graph.py:32]: Conv2DTranspose?

File: model_compression_toolkit/core/pytorch/statistics_correction/pytorch_compute_activation_bias_correction_of_graph.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    # Match bypass layers what don't affect the quantization process.
+    bypass_node = NodeOperationMatcher(reshape) | \
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/statistics_correction/pytorch_compute_activation_bias_correction_of_graph.py:35]: Can you take this list from the TPC's no_quant and quant_preserving list?

File: tests/keras_tests/feature_networks_tests/feature_networks/activation_bias_correction_test.py
Comment (elad-c) [inline tests/keras_tests/feature_networks_tests/feature_networks/activation_bias_correction_test.py:1]: This file has too much duplications. You can switch to a single test that takes the linear layer type and bypass layers as inputs. also the threshold.
and don't call all test "Base", only common classes that other tests inherit from.

File: tests/pytorch_tests/model_tests/feature_models/activation_bias_correction_test.py
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/activation_bias_correction_test.py:1]: Same as keras file comment

**Review by elad-c — CHANGES_REQUESTED** [review]

File: tests/pytorch_tests/model_tests/feature_models/activation_bias_correction_test.py [hunk 11]
```diff
@@ -86,74 +87,39 @@ def forward(self, x):
...
+                                              f"value in case of activation_bias_correction_threshold "
+                                              f"{self.activation_bias_correction_threshold}.")
+            else:
+                self.unit_test.assertFalse(np.array_equal(bias, bias_after_activation_bias_correction),
+                                           msg=f"Error in activation bias correction: expected a change in the bias "
+                                               f"value.")
+        else:
+            self.unit_test.assertTrue(np.array_equal(bias, bias_after_activation_bias_correction),
+                                      msg=f"Error in activation bias correction: expected no change in the bias value "
+                                          f"in case of conv with kernel different than 1 or (1, 1).")
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/activation_bias_correction_test.py:125]: I think you should add an output compare, so that we make sure that the quantized model is built correctly.

File: tests/keras_tests/feature_networks_tests/feature_networks/activation_bias_correction_test.py [hunk 12]
```diff
@@ -27,171 +28,58 @@
...
+                                              f"value in case of activation_bias_correction_threshold "
+                                              f"{self.activation_bias_correction_threshold}.")
+            else:
+                self.unit_test.assertFalse(np.array_equal(bias, bias_after_activation_bias_correction),
+                                           msg=f"Error in activation bias correction: expected a change in the bias "
+                                               f"value.")
+        else:
+            self.unit_test.assertTrue(np.array_equal(bias, bias_after_activation_bias_correction),
+                                      msg=f"Error in activation bias correction: expected no change in the bias value "
+                                          f"in case of conv with kernel 2.")
```
Comment (elad-c) [inline tests/keras_tests/feature_networks_tests/feature_networks/activation_bias_correction_test.py:85]: I think you should add an output compare, so that we make sure that the quantized model is built correctly.

**Review by elad-c — CHANGES_REQUESTED** [review]

**Review by elad-c — APPROVED** [review]
